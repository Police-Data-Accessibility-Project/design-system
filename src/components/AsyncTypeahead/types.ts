export interface PdapAsyncTypeaheadProps<T> {
	id: string;
	placeholder?: string;
	items: T[] | undefined;
	formatItemForDisplay?: (item: T) => string;
	error?: string;
	position?: 'top' | 'bottom';
}
