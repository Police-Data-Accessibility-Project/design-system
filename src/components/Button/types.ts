export interface PdapButtonProps {
	intent?: 'primary' | 'secondary' | 'tertiary';
	isLoading?: boolean;
}
