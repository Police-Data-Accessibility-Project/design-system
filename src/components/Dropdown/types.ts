export enum PdapDropdownTriggerType {
	PRESS = 'press',
	HOVER = 'hover',
}

export interface PdapDropdownProps {
	defaultOpen?: boolean;
	disabled?: boolean;
	triggerOn?: PdapDropdownTriggerType;
}
