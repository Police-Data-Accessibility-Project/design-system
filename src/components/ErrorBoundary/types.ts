import { ComponentPublicInstance } from 'vue';

export interface PdapErrorEmitted {
	error: Error;
	vm: ComponentPublicInstance | null;
	info?: string;
}

export interface PdapErrorBoundaryProps {
	component: string;
	onError?: (
		error: Error,
		target?: ComponentPublicInstance | null | undefined,
		info?: string
	) => void;
	params?: Record<string, string>;
}
