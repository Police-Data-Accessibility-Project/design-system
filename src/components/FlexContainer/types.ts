export interface PdapFlexContainerProps {
	alignment?: 'center' | 'start';
	className?: string;
	component?: string;
}
