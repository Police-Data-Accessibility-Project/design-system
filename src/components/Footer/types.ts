import { FOOTER_LINK_ICONS } from './constants';

export interface PdapFooterSocialLinks {
	href?: string;
	path?: string;
	text: string;
	icon?: FooterIconName;
}
export type FooterIconName =
	(typeof FOOTER_LINK_ICONS)[keyof typeof FOOTER_LINK_ICONS];

export interface PdapFooterProps {
	fundraisingData: {
		raised: number;
		goal: number;
	};
}
