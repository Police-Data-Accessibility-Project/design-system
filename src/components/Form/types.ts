import { PdapInputCheckboxProps, PdapInputTextProps } from '../Input/types';

export type PdapLengthRules = 'maxLength' | 'minLength';

export interface PdapFormValidator<T> {
	message?: string;
	value: T;
}

/**
 * Keyed by currently used validators.
 * Add any Vuelidate validators or custom ones here as we need them.
 * See https://vuelidate-next.netlify.app/validators.html#using-builtin-validators for more.
 *
 */
export interface PdapFormValidators {
	maxLength: PdapFormValidator<number>;
	minLength: PdapFormValidator<number>;
	required: PdapFormValidator<boolean>;
}

export interface PdapFormSchemaEntryInputCheckbox
	extends PdapInputCheckboxProps {
	validators?: Partial<PdapFormValidators>;
}
export interface PdapFormSchemaEntryInputText extends PdapInputTextProps {
	validators?: Partial<PdapFormValidators>;
}
export type PdapFormSchemaEntry =
	| PdapFormSchemaEntryInputCheckbox
	| PdapFormSchemaEntryInputText;
export type PdapFormSchema = PdapFormSchemaEntry[];
export type PdapFormData = Record<string, unknown>;

/**
 * PDAP Form props interface.
 */
export interface PdapFormProps {
	error?: string | undefined | null;
	id: string;
	name: string;
	schema: PdapFormSchema;
}
