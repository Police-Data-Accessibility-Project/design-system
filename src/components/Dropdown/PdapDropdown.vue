<template>
	<div
		ref="dropdownRef"
		v-click-outside="closeDropdown"
		class="pdap-dropdown"
		@keydown.escape="closeDropdown"
		v-on="dropdownHandlers"
	>
		<Button
			:aria-controls="dropdownId"
			:aria-expanded="open"
			:disabled="disabled"
			aria-haspopup="true"
			intent="tertiary"
			:class="{
				'pdap-dropdown-trigger': true,
				'pointer-events-none': disabled,
				'cursor-not-allowed': disabled,
			}"
			data-test="dropdown-trigger"
			@click.prevent.stop="press"
			@keydown.enter.space.prevent.stop="press"
		>
			<slot name="trigger" />
		</Button>

		<Transition name="dropdown" appear>
			<div
				v-show="open"
				:id="dropdownId"
				ref="contentRef"
				:aria-hidden="!open"
				class="pdap-dropdown-content"
				data-test="dropdown-content"
				role="menu"
			>
				<slot name="content" />
			</div>
		</Transition>
	</div>
</template>

<script setup lang="ts">
import { computed, onMounted, ref } from 'vue';
import { PdapDropdownProps, PdapDropdownTriggerType } from './types';
import { default as vClickOutside } from '../../directives/click-outside';
import { Button } from '../Button';

// Props
const props = withDefaults(defineProps<PdapDropdownProps>(), {
	defaultOpen: false,
	disabled: false,
	triggerOn: PdapDropdownTriggerType.PRESS,
});

// Emits
const emit = defineEmits(['update:open', 'press']);

// Vars
const dropdownId = 'pdap-dropdown';
const contentRef = ref<HTMLElement | null>(null);
const dropdownRef = ref<HTMLElement | null>(null);
const open = ref<boolean>(props.defaultOpen);
const triggerOnHover = computed(
	() => props.triggerOn === PdapDropdownTriggerType.HOVER
);

function press(event: Event) {
	emit('press', event);
	if (props.triggerOn === PdapDropdownTriggerType.PRESS) toggleDropdown();
}

const toggleDropdown = () => {
	if (!props.disabled) {
		open.value = !open.value;
		emit('update:open', open.value);
	}
};

function openDropdown() {
	open.value = true;
	emit('update:open', true);
}

function closeDropdown() {
	open.value = false;
	emit('update:open', false);
}

function getMaxHeight() {
	if (!contentRef.value) return;
	contentRef.value.style.display = 'flex';
	contentRef.value.style.setProperty('max-height', 'max-content');
	const height = contentRef.value.scrollHeight;
	contentRef.value.style.display = 'none';
	contentRef.value.style.removeProperty('max-height');
	const root = document?.querySelector(':root') as HTMLElement;

	const previousMaxHeight = root?.style?.getPropertyValue(
		'--dropdown-content-max-height'
	);

	// If current dropdown shorter than current max height var, return
	if (Number(previousMaxHeight?.match(/\d+/)?.join('')) >= height) return;

	// Otherwise, increase var
	if (contentRef.value && root)
		root.style.setProperty('--dropdown-content-max-height', `${height + 16}px`);
}

const dropdownHandlers = {
	mouseenter: triggerOnHover.value ? openDropdown : undefined,
	focus: triggerOnHover.value ? openDropdown : undefined,
	focusin: triggerOnHover.value ? openDropdown : undefined,
	focusout: triggerOnHover.value ? closeDropdown : undefined,
	mouseleave: triggerOnHover.value ? closeDropdown : undefined,
	blur: triggerOnHover.value ? closeDropdown : undefined,
};

onMounted(() => {
	getMaxHeight();
});
</script>

<script lang="ts">
/**
 * # `Dropdown`
 * An accessible dropdown menu that can be triggered by click or hover and positioned in various placements. It provides keyboard accessibility features such as toggling the dropdown on enter/space and closing on escape.
 *
 * ## Props
 * @prop {boolean} defaultOpen Controls the initial `isOpen` state
 * @prop {boolean} disabled Disabled state (forwarded to trigger button)
 * @prop {PdapDropdownTriggerType} triggerType `'press' | 'hover'`
 *
 */
export default {
	name: 'PdapButton',
};
</script>

<style>
@tailwind components;

:root {
	--dropdown-content-max-height: 0px;
}

@layer components {
	.pdap-dropdown {
		@apply relative inline-block w-max;
	}

	.pdap-dropdown-trigger {
		@apply cursor-pointer p-0 px-2;
	}

	.pdap-dropdown-content {
		@apply flex flex-col w-full overflow-hidden relative max-h-[var(--dropdown-content-max-height)] px-2;
	}

	.pdap-dropdown-content.dropdown-enter-active,
	.pdap-dropdown-content.dropdown-leave-active {
		transition:
			opacity 0.5s ease-in,
			max-height 0.7s ease;
	}

	.pdap-dropdown-content.dropdown-enter-from,
	.pdap-dropdown-content.dropdown-leave-to {
		@apply max-h-0 opacity-0;
	}

	.pdap-dropdown-content > * {
		@apply my-2 p-0 w-full max-w-[calc(100%-4px)] items-start text-start;
	}
}
</style>
