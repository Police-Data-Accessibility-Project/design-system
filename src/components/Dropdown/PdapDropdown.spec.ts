import { describe, it, expect, beforeEach } from 'vitest';
import { shallowMount } from '@vue/test-utils';
import PdapDropdown from './PdapDropdown.vue';
import { PdapDropdownTriggerType } from './types';

let wrapper;
let trigger;
let content;

describe('PdapDropdown', () => {
	beforeEach(() => {
		wrapper = shallowMount(PdapDropdown, {
			slots: {
				trigger: 'Toggle Dropdown',
				content: '<ul><li>Option 1</li><li>Option 2</li><li>Option 3</li></ul>',
			},
			attachTo: document.body,
		});

		trigger = wrapper.find('[data-test="dropdown-trigger"]');
		content = wrapper.find('[data-test="dropdown-content"]');
	});

	it('should open and close dropdown correctly', async () => {
		expect(content.isVisible()).toBe(false);

		await trigger.trigger('click');
		expect(content.isVisible()).toBe(true);

		await trigger.trigger('click');
		expect(content.isVisible()).toBe(false);
		expect(wrapper.html()).toMatchSnapshot();
	});

	it('should be accessible with keyboard', async () => {
		expect(content.isVisible()).toBe(false);

		await trigger.trigger('keydown.enter');
		expect(content.isVisible()).toBe(true);

		await trigger.trigger('keydown.escape');
		expect(content.isVisible()).toBe(false);

		expect(wrapper.html()).toMatchSnapshot();
	});

	it('should respect disabled prop', async () => {
		wrapper = shallowMount(PdapDropdown, {
			props: {
				disabled: true,
			},
			slots: {
				trigger: 'Toggle Dropdown',
				content: '<ul><li>Option 1</li><li>Option 2</li><li>Option 3</li></ul>',
			},
			attachTo: document.body,
		});
		trigger = wrapper.find('[data-test="dropdown-trigger"]');
		content = wrapper.find('[data-test="dropdown-content"]');

		await trigger.trigger('click');
		expect(content.isVisible()).toBe(false);

		expect(wrapper.html()).toMatchSnapshot();
	});

	it('should respect the triggerOn prop', async () => {
		wrapper = shallowMount(PdapDropdown, {
			props: {
				triggerOn: PdapDropdownTriggerType.HOVER,
			},
			slots: {
				trigger: 'Toggle Dropdown',
				content: '<ul><li>Option 1</li><li>Option 2</li><li>Option 3</li></ul>',
			},
			attachTo: document.body,
		});
		trigger = wrapper.find('[data-test="dropdown-trigger"]');
		content = wrapper.find('[data-test="dropdown-content"]');

		await trigger.trigger('focusin');
		expect(content.isVisible()).toBe(true);

		await trigger.trigger('focusout');
		expect(content.isVisible()).toBe(false);

		expect(wrapper.html()).toMatchSnapshot();
	});
});
