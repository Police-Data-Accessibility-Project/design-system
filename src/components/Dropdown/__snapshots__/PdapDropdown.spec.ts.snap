// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PdapDropdown > should be accessible with keyboard 1`] = `
<div class="pdap-dropdown" role="presentation">
  <pdap-button-stub aria-controls="test-dropdown" aria-expanded="false" aria-haspopup="true" class="pdap-dropdown-trigger" disabled="false" intent="tertiary" isloading="false"></pdap-button-stub>
  <transition-stub appear="true" css="true" name="dropdown" persisted="true">
    <div aria-hidden="true" class="pdap-dropdown-content" id="test-dropdown" role="menu" style="display: none;">
      <ul>
        <li>Option 1</li>
        <li>Option 2</li>
        <li>Option 3</li>
      </ul>
    </div>
  </transition-stub>
</div>
`;

exports[`PdapDropdown > should open and close dropdown correctly 1`] = `
<div class="pdap-dropdown" role="presentation">
  <pdap-button-stub aria-controls="test-dropdown" aria-expanded="false" aria-haspopup="true" class="pdap-dropdown-trigger" disabled="false" intent="tertiary" isloading="false"></pdap-button-stub>
  <transition-stub appear="true" css="true" name="dropdown" persisted="true">
    <div aria-hidden="true" class="pdap-dropdown-content" id="test-dropdown" role="menu" style="display: none;">
      <ul>
        <li>Option 1</li>
        <li>Option 2</li>
        <li>Option 3</li>
      </ul>
    </div>
  </transition-stub>
</div>
`;

exports[`PdapDropdown > should respect disabled prop 1`] = `
<div class="pdap-dropdown" role="presentation">
  <pdap-button-stub aria-controls="test-dropdown" aria-expanded="false" aria-haspopup="true" class="pdap-dropdown-trigger pointer-events-none cursor-not-allowed" disabled="true" intent="tertiary" isloading="false"></pdap-button-stub>
  <transition-stub appear="true" css="true" name="dropdown" persisted="true">
    <div aria-hidden="true" class="pdap-dropdown-content" id="test-dropdown" role="menu" style="display: none;">
      <ul>
        <li>Option 1</li>
        <li>Option 2</li>
        <li>Option 3</li>
      </ul>
    </div>
  </transition-stub>
</div>
`;

exports[`PdapDropdown > should respect the triggerOn prop 1`] = `
<div class="pdap-dropdown" role="presentation">
  <pdap-button-stub aria-controls="test-dropdown" aria-expanded="false" aria-haspopup="true" class="pdap-dropdown-trigger" disabled="false" intent="tertiary" isloading="false"></pdap-button-stub>
  <transition-stub appear="true" css="true" name="dropdown" persisted="true">
    <div aria-hidden="true" class="pdap-dropdown-content" id="test-dropdown" role="menu" style="display: none;">
      <ul>
        <li>Option 1</li>
        <li>Option 2</li>
        <li>Option 3</li>
      </ul>
    </div>
  </transition-stub>
</div>
`;
